Simplified the calculation of goal direction.

lewie-donckers opened the pull request:
Simplified the calculation of the direction of the goal in `Controller::update()`. Fix for #82 .

lewie-donckers commented:
@Rayman @rokusottervanger @Timple 
I need some help. I noticed that I introduced a change in behavior which can be seen when running the visualization of the tests. I tracked it down to this commit.

In certain tests, you can see that the robot takes longer to stop and turns a bit more at the end of the test.

Example before change:
![image](https://user-images.githubusercontent.com/20455146/156151879-21e875d8-a5d8-480c-a150-d93ec9172c34.png)

Example after change:
![image](https://user-images.githubusercontent.com/20455146/156152545-88223d43-1a4f-4c2a-a59a-0faa9ab2bff5.png)

Any idea what's going on and how to fix it?

rokusottervanger commented:
I'm not sure. But could you maybe refactor this first, so that this calculation is isolated as a separate function. Then a unit test can show if/how the behavior changes.

rokusottervanger commented on src/controller.cpp:
```diff
@@ -529,14 +522,13 @@ Controller::UpdateResult Controller::update(
   // However, if robot is not facing to the same direction as the local velocity target vector, don't enable end_phase.
   // This is to avoid skipping paths that start with opposite velocity.
-  if ((distance_to_goal_ <= fabs(d_end_phase)) && have_same_sign(target_x_vel, angle_to_goal)) {
+  if ((distance_to_goal_ <= fabs(d_end_phase)) && in_direction_of_goal) {
```
Where's the comparison with the target_x_vel gone?

rokusottervanger commented on src/controller.cpp:
```diff
@@ -529,14 +522,13 @@ Controller::UpdateResult Controller::update(
-  const double angle_to_goal = atan2(base_to_goal.getOrigin().x(), -base_to_goal.getOrigin().y());
+  const auto in_direction_of_goal =
+    !std::signbit(current_tf.getOrigin().dot(current_goal_.getOrigin()));
 
```
I like the seeming simplicity of this, but `current_tf` and `current_goal_` are both defined in the `map` frame, so you're now projecting the vector from `map` to `current_tf` onto the vector from `map` to `current_goal_`.  This doesn't do the same as in the original calculation.

rokusottervanger commented on src/controller.cpp:
```diff
@@ -529,14 +522,13 @@ Controller::UpdateResult Controller::update(
-  // Get 'angle' towards current_goal
-  tf2::Transform base_to_goal = current_tf.inverseTimes(current_goal_);
-  const double angle_to_goal = atan2(base_to_goal.getOrigin().x(), -base_to_goal.getOrigin().y());
+  const auto in_direction_of_goal =
+    !std::signbit(current_tf.getOrigin().dot(current_goal_.getOrigin()));
```
I was thinking something along the lines of this:
```suggestion
const auto base_to_next_path_pose = current_goal_.getOrigin() - current_tf.getOrigin();
  const auto in_direction_of_goal =
    !std::signbit(current_goal_.getBasis().tdotx(base_to_next_path_pose)*target_x_vel);
```
But I'm not sure if it's entirely correct. I now project onto the current_goal_ x vector, but one could also pick the current_tf x vector. Not sure which is correct, and how to test this in an integration test.

Also, I was inspired by your `!std::signbit`, and added the multiplication with the current velocity setpoint. That doesn't make it any more transparent...

lewie-donckers replied:
I'm not sure if it's correct or not but it did give different results than the original implementation (for certain inputs). Based on your remarks, I changed it to the following: 

```cpp
  const auto base_to_next_path_pose = current_goal_.getOrigin() - current_tf.getOrigin();
  const auto in_direction_of_goal =
    !std::signbit(current_tf.getBasis().tdotx(base_to_next_path_pose) * target_x_vel);
```

This does give the same results as the original implementation for all inputs that occur in the automated tests.

I plan to do the following:
- move the new implementation to a named function in `calculations.h/cpp`
- add some unit tests

I could use some help with the documentation and coming up with interesting inputs for the unit tests

lewie-donckers commented:
@Timple @rokusottervanger @Rayman 

I have reworked this PR. I fixed the new implementation with the help of @rokusottervanger . The calculation is now moved to a separate function and I've added some unit tests.

I tested this by running a build with both the old and new calculation in tandem. It logged if they were not the same. None of the automated tests showed any differences.
